### PR TITLE
Fix flacky test introduced by af2f343648f4cb3deca5314174e0e579f9fec846

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

af2f343648f4cb3deca5314174e0e579f9fec846 introduced a test-case which was flacky due of multiple problems:

- we called writeAndFlush(...) in channelRead(...) and assumed it will only be called once. This is true most of the times but it may be called multile times if the data is fragemented.
- we didnt guard against the possibility that channelRead(...) is called with an empty buffer

Modifications:

- Call writeAndFlush(...) in channelActive(...) so we are sure its only called once and close the channel once we wrote the data
- only compare the data after we received a close so we are sure there isnt anything extra received
- check for exception and if we catched one fail the test.

Result:

No flacky test anymore and easier to debug issues that accour because of a catched exception.

### CODE REVIEW

**Comment by normanmaurer** [comment]
Let me merge this to unblock the release.

4.1 (661bd8682913e62a25a8952362daff60fecf12bf)

**Comment by normanmaurer** [comment]
For the record for example this exception was sometimes in the logs before:

```
09:38:09.284 [Time-limited test] INFO  i.n.t.t.s.CompositeBufferGatheringWriteTest - Running: testSingleCompositeBufferWrite 2 of 3 (ServerBootstrap(ServerBootstrapConfig(group: NioEventLoopGroup, channelFactory: NioServerSocketChannel.class, localAddress: /0:0:0:0:0:0:0:1:0, options: {ALLOCATOR=PooledByteBufAllocator(directByDefault: true)}, childGroup: NioEventLoopGroup, childOptions: {ALLOCATOR=PooledByteBufAllocator(directByDefault: true)})) + Bootstrap(BootstrapConfig(group: OioEventLoopGroup, channelFactory: OioSocketChannel.class, options: {SO_TIMEOUT=10, ALLOCATOR=PooledByteBufAllocator(directByDefault: true)}, resolver: io.netty.resolver.DefaultAddressResolverGroup@1d3ad8ab))) with PooledByteBufAllocator
09:38:09.298 [testsuite-oio-worker-5-1] WARN  i.n.channel.DefaultChannelPipeline - An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
io.netty.util.IllegalReferenceCountException: refCnt: 0
	at io.netty.buffer.AbstractByteBuf.ensureAccessible(AbstractByteBuf.java:1415)
	at io.netty.buffer.AbstractByteBuf.ensureWritable0(AbstractByteBuf.java:270)
	at io.netty.buffer.AbstractByteBuf.ensureWritable(AbstractByteBuf.java:265)
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1079)
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1072)
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1062)
	at io.netty.testsuite.transport.socket.CompositeBufferGatheringWriteTest$2$1.channelRead(CompositeBufferGatheringWriteTest.java:75)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:935)
	at io.netty.channel.oio.AbstractOioByteChannel.doRead(AbstractOioByteChannel.java:166)
	at io.netty.channel.oio.AbstractOioChannel$1.run(AbstractOioChannel.java:37)
	at io.netty.channel.ThreadPerChannelEventLoop.run(ThreadPerChannelEventLoop.java:67)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
09:38:09.299 [Time-limited test] INFO  i.n.t.t.s.CompositeBufferGatheringWriteTest - Running: testSingleCompositeBufferWrite 3 of 3 (ServerBootstrap(ServerBootstrapConfig(group: OioEventLoopGroup, channelFactory: OioServerSocketChannel.class, localAddress: /0:0:0:0:0:0:0:1:0, options: {SO_TIMEOUT=10, ALLOCATOR=PooledByteBufAllocator(directByDefault: true)}, childGroup: OioEventLoopGroup, childOptions: {ALLOCATOR=PooledByteBufAllocator(directByDefault: true)})) + Bootstrap(BootstrapConfig(group: NioEventLoopGroup, channelFactory: NioSocketChannel.class, options: {ALLOCATOR=PooledByteBufAllocator(directByDefault: true)}, resolver: io.netty.resolver.DefaultAddressResolverGroup@1d3ad8ab))) with PooledByteBufAllocator
```

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java [hunk 1]
```diff
@@ -73,16 +75,35 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) {
...
+                            // IOException is fine as it will also close the channel and may just be a connection reset.
+                            if (!(cause instanceof IOException)) {
+                                clientReceived.set(cause);
+                                latch.countDown();
+                            }
+                        }
+
+                        @Override
+                        public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+                            if (clientReceived.compareAndSet(null, aggregator)) {
```
Comment (Scottmitch) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java]: do we need to worry about channel inactive being called multiple times?

Comment (normanmaurer) [reply testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java]: @Scottmitch no really but I wanted to ensure we not override the exception that we may have set in `exceptionCaught(...)` before.

**Pull request closed without merging**